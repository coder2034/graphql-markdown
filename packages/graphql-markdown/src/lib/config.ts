import {
  GraphQLExtensionDeclaration,
  GraphQLProjectConfig,
  loadConfig,
} from "graphql-config";
import { DocumentNode } from "graphql";

import { GraphQLFileLoader } from "@graphql-tools/graphql-file-loader";
import { JsonFileLoader } from "@graphql-tools/json-file-loader";
import { UrlLoader } from "@graphql-tools/url-loader";

import { ExtensionAPI, LoadConfigOptions, Maybe } from "..";

const EXTENSION_NAME = "graphql-markdown" as const;

export const OPTION = {
  LAYOUTS: "layouts",
  MDX: "mdx",
  OUTPUT: "output",
};

export type ConfigurationOptions = {
  excludes?: readonly string[];
  layouts?: string;
  mdx?: boolean;
  output?: string;
};

const defaultOptions: ConfigurationOptions = {
  excludes: [] as const,
  [OPTION.LAYOUTS]: "./layouts" as const,
  [OPTION.OUTPUT]: false as const,
  output: "./output" as const,
} as const;

const setFileLoaderExtension: GraphQLExtensionDeclaration = (
  api: ExtensionAPI
) => {
  [new GraphQLFileLoader(), new JsonFileLoader(), new UrlLoader()].forEach(
    (loader) => {
      return api.loaders.schema.register(loader);
    }
  );
  return { name: EXTENSION_NAME };
};

const loadConfiguration = async (
  options?: LoadConfigOptions
): Promise<Maybe<GraphQLProjectConfig>> => {
  const config = await loadConfig({
    ...options,
    extensions: [setFileLoaderExtension],
  });
  return config?.getDefault();
};

export class Configuration {
  static project: Maybe<GraphQLProjectConfig>;
  static extension: ConfigurationOptions;

  static load = async (options?: LoadConfigOptions): Promise<void> => {
    const configuration = await loadConfiguration(options);
    Configuration.project = configuration;
    Configuration.extension = {
      ...defaultOptions,
      ...Configuration.project?.extension(EXTENSION_NAME),
    };
  };

  static get = (name: string): string => {
    return Configuration.extension[name];
  };

  static set = (name: string, value: string): void => {
    Configuration.extension[name] = value;
  };

  // eslint-disable-next-line require-await
  static schema = async (): Promise<Maybe<DocumentNode>> => {
    return Configuration.project!.getSchema("DocumentNode");
  };
}
