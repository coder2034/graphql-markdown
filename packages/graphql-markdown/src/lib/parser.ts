import {
  ASTNode,
  ArgumentNode,
  BooleanValueNode,
  DirectiveDefinitionNode,
  DirectiveNode,
  DocumentNode,
  EnumValueDefinitionNode,
  EnumValueNode,
  FieldDefinitionNode,
  FloatValueNode,
  InputValueDefinitionNode,
  IntValueNode,
  ListTypeNode,
  ListValueNode,
  NameNode,
  NamedTypeNode,
  NonNullTypeNode,
  ObjectFieldNode,
  ObjectTypeDefinitionNode,
  ObjectValueNode,
  OperationTypeDefinitionNode,
  OperationTypeNode,
  StringValueNode,
  TypeDefinitionNode,
  visit,
} from "graphql";

import { ParsedNode } from "..";

const GRAPHQL_KIND = {
  DirectiveDefinition: "DirectiveDefinition",
  EnumTypeDefinition: "EnumTypeDefinition",
  InputObjectTypeDefinition: "InputObjectTypeDefinition",
  InputTypeDefinition: "InputTypeDefinition",
  InterfaceTypeDefinition: "InterfaceTypeDefinition",
  ObjectTypeDefinition: "ObjectTypeDefinition",
  OperationTypeDefinition: "OperationTypeDefinition",
  ScalarTypeDefinition: "ScalarTypeDefinition",
  UnionTypeDefinition: "UnionTypeDefinition",
} as const;

const SIMPLIFIED_NODE_KIND = {
  DirectiveDefinition: "directive",
  EnumTypeDefinition: "enum",
  InputObjectTypeDefinition: "input",
  InputTypeDefinition: "interface",
  Mutation: "mutation",
  ObjectTypeDefinition: "object",
  OperationTypeDefinition: "operation",
  Query: "query",
  ScalarTypeDefinition: "scalar",
  Subscription: "subscription",
  UnionTypeDefinition: "union",
} as const;

const OPERATION_TYPE = {
  MUTATION: "Mutation",
  QUERY: "Query",
  SUBSCRIPTION: "Subscription",
} as const;

const visitor = {
  Argument: (node: ArgumentNode) => {
    return node;
  },
  BooleanValue: ({ value }: BooleanValueNode) => {
    return value;
  },
  Directive: (node: DirectiveNode) => {
    return { ...node, name: `@${node.name}` };
  },
  DirectiveDefinition: (node: DirectiveDefinitionNode) => {
    return { ...node, name: `@${node.name}` };
  },
  Document: ({ definitions }: DocumentNode) => {
    return definitions;
  },
  EnumValue: ({ value }: EnumValueNode) => {
    return value;
  },
  EnumValueDefinition: (node: EnumValueDefinitionNode) => {
    return node;
  },
  FieldDefinition: ({
    type,
    name,
    arguments: args,
    directives,
    kind,
    description,
  }: FieldDefinitionNode) => {
    return {
      ...type,
      arguments: args,
      description,
      directives,
      kind,
      name,
    };
  },
  FloatValue: ({ value }: FloatValueNode) => {
    return value;
  },
  InputValueDefinition: ({
    type,
    name,
    directives,
    defaultValue,
    kind,
    description,
  }: InputValueDefinitionNode) => {
    return {
      ...type,
      defaultValue,
      description,
      directives,
      kind,
      name,
    };
  },
  IntValue: ({ value }: IntValueNode) => {
    return value;
  },
  ListType: ({ type }: ListTypeNode) => {
    return {
      ...type,
      isList: true,
    };
  },
  ListValue: ({ values }: ListValueNode) => {
    return values;
  },
  Name: ({ value }: NameNode) => {
    return value;
  },
  NamedType: ({ name }: NamedTypeNode) => {
    return {
      isList: false,
      isNull: true,
      type: name,
    };
  },
  NonNullType: ({ type }: NonNullTypeNode) => {
    return {
      ...type,
      isNull: false,
    };
  },
  NullValue: () => {
    return "null";
  },
  ObjectField: (node: ObjectFieldNode) => {
    return node;
  },
  ObjectValue: ({ fields }: ObjectValueNode) => {
    return fields;
  },
  OperationType: (node: OperationTypeNode) => {
    return node.toString();
  },
  OperationTypeDefinition: (node: OperationTypeDefinitionNode) => {
    return node;
  },
  StringValue: ({ value }: StringValueNode) => {
    return value;
  },
  TypeDefinition: (node: TypeDefinitionNode) => {
    return node;
  },
} as const;

const OperationTypes: readonly string[] = [
  OPERATION_TYPE.QUERY,
  OPERATION_TYPE.MUTATION,
  OPERATION_TYPE.SUBSCRIPTION,
] as const;

const isNodeTypeOperation = (
  node: ASTNode | ParsedNode
): node is ObjectTypeDefinitionNode => {
  return (
    "name" in node &&
    typeof node.name !== "undefined" &&
    OperationTypes.includes(node.name.toString()) &&
    "fields" in node
  );
};

export const getSimplifiedNodeKind = (node: ASTNode | ParsedNode): string => {
  const kind = "operation" in node ? node.operation : node.kind;
  return SIMPLIFIED_NODE_KIND[kind] ?? node.kind;
};

export const parseSchema = (schema: DocumentNode): ParsedNode[] => {
  const nodes: ASTNode[] = visit(schema, { leave: visitor });
  return nodes.flatMap((node) => {
    if (isNodeTypeOperation(node) && typeof node.fields !== "undefined") {
      return node.fields.map<ParsedNode>((operation: FieldDefinitionNode) => {
        const parsedNode = {
          ...operation,
          kind: GRAPHQL_KIND.OperationTypeDefinition,
          operation: node.name,
        } as unknown as ParsedNode;
        return {
          ...parsedNode,
          simplifiedKind: getSimplifiedNodeKind(parsedNode),
        } as ParsedNode;
      });
    }
    return {
      ...node,
      simplifiedKind: getSimplifiedNodeKind(node),
    } as ParsedNode;
  });
};
