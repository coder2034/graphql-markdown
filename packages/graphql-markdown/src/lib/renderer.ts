import * as path from "path";

import * as Eta from "eta";
import * as prettier from "prettier";

import { __basedir, getConfigurationOption } from "../";

export const renderNode = async (node: any): Promise<string> => {
  const layout = path.resolve(
    __basedir,
    getConfigurationOption("layouts"),
    "layout"
  );
  const result = (await Eta.renderFile(layout, node)) as string;

  return prettier.format(result, { parser: "markdown" });
};
