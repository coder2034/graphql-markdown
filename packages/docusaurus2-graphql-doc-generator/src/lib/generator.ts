import { promises as fs } from "fs";
import path from "path";

import prettier from "prettier";

import { generateMarkdownFromSchema } from "@edno/graphql-markdown";

import { PluginOptions, Sidebar } from "../types";

const SIDEBAR_FILE = "sidebar-schema.js";

const prettifyJSON = (content: string): string => {
  return prettier.format(content, { parser: "json" });
};

const prettifyMarkdown = (content: string): string => {
  return prettier.format(content, { parser: "markdown" });
};

const renderHomepage = async (
  homepageLocation: string,
  options: PluginOptions
): Promise<string> => {
  const homepage = await fs.readFile(homepageLocation, "utf8");

  const filePath = path.join(options.rootPath, path.basename(homepageLocation));
  await fs.writeFile(filePath, prettifyMarkdown(homepage), "utf8");

  return path.relative("./", filePath);
};

const renderSidebar = async ({
  rootPath,
  baseURL,
}: PluginOptions): Promise<string> => {
  const sidebar: Record<string, Sidebar> = {
    schemaSidebar: [
      {
        dirName: baseURL,
        type: "autogenerated",
      },
    ],
  };

  const filePath = path.join(rootPath, SIDEBAR_FILE);
  await fs.writeFile(filePath, prettifyJSON(JSON.stringify(sidebar)), "utf8");

  return path.relative("./", filePath);
};

export const generateDocFromSchema = async (
  schemaLocation: string,
  options: PluginOptions
): Promise<{ pages: number; sidebar: string }> => {
  const pages = await generateMarkdownFromSchema({ saveToFiles: true });

  // eslint-disable-next-line no-magic-numbers
  if (typeof pages === "undefined" || pages.length === 0) {
    throw new Error("Parsing error, no pages generated");
  }

  if (typeof options === "undefined") {
    throw new Error("Parsing error, no pages generated");
  }

  await renderHomepage(options.homepage, options);
  const sidebarPath = await renderSidebar(options);

  return { pages: pages.length, sidebar: sidebarPath };
};
