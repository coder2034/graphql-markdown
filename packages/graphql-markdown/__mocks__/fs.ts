import { IFS } from "unionfs/lib/fs";
import { Volume } from "memfs";
import { ufs } from "unionfs";

const fs = jest.requireActual("fs");

const vol = Volume.fromJSON({});
vol.mkdirSync(process.cwd(), { recursive: true });

module.exports = ufs.use(fs).use(vol as unknown as IFS);
