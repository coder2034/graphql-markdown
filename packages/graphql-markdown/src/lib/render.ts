import * as path from "path";
import { promises as fs } from "fs";

import * as Eta from "eta";
import * as prettier from "prettier";
import slugify from "slugify";
import kebabCase from "kebab-case";

import { __basedir, getConfigurationOption, getSimplifiedNodeKind } from "..";

export const renderNode = async (node: any): Promise<string> => {
  const layout = path.resolve(
    __basedir,
    getConfigurationOption("layouts"),
    "layout"
  );
  const result = (await Eta.renderFile(layout, node)) as string;

  return prettier.format(result, { parser: "markdown" });
};

export const slug = (input) => slugify(kebabCase(input));

export const saveMarkdownFile = async (node: any): Promise<string> => {
  const folder = getSimplifiedNodeKind(node);
  const filename = `${slug(node.name)}.md`;
  const filePath = path.resolve(
    __basedir,
    getConfigurationOption("output"),
    slug(folder),
    filename
  );

  const markdown = await renderNode(node);
  await fs.writeFile(filePath, markdown);

  return filePath;
};
