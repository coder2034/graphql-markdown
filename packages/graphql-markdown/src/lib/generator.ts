import { Configuration, ConfigurationOptions } from "./config";
import { LoadConfigOptions, ParsedNode, Result } from "..";
import { renderNode, saveMarkdownFile } from "./render";
import { parseSchema } from "./parser";

const loadNodes = async (
  options?: LoadConfigOptions
): Promise<ParsedNode[]> => {
  await Configuration.load(options);

  const schema = await Configuration.schema();

  if (typeof schema === "undefined") {
    throw new Error("An error occurred while loading the schema");
  }

  return parseSchema(schema);
};

type Options = {
  saveToFiles?: boolean;
  loaderOptions?: LoadConfigOptions;
  generatorOptions?: ConfigurationOptions;
};

export const generateMarkdownFromSchema = async (
  options?: Options
): Promise<Result[]> => {
  const nodes = await loadNodes(options?.loaderOptions);

  const processor =
    options?.saveToFiles === true ? saveMarkdownFile : renderNode;

  const results = await Promise.all(
    nodes.map(async (node: ParsedNode): Promise<Result> => {
      const result = await processor(node, options?.generatorOptions);
      return {
        name: node.name,
        type: node.type,
        /* eslint-disable prettier/prettier */
        ...(options?.saveToFiles === true
          ? { filepath: result }
          : { markdown: result }),
        /* eslint-enable */
      };
    })
  );

  return results;
};
