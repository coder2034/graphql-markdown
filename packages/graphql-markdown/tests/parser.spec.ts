import * as path from "path";
import { promises as fs } from "fs";

import { ASTNode, DocumentNode } from "graphql";

import { getSimplifiedNodeKind, parseSchema } from "../src/lib/parser";

describe("parser", () => {
  describe("getSimplifiedNodeKind", () => {
    const operations = [
      ["./__data__/node/mutation", "mutation"],
      ["./__data__/node/query", "query"],
      ["./__data__/node/subscription", "subscription"],
    ];
    it.each(operations)(
      "should return the operation object name as simplified type",
      async (source: string, expected: string) => {
        expect.hasAssertions();

        const node: ASTNode = (await import(source)).default as ASTNode;
        expect(getSimplifiedNodeKind(node)).toEqual(expected);
      }
    );
  });

  describe("parseSchema", () => {
    let schema: DocumentNode; // eslint-disable-line init-declarations

    beforeAll(async () => {
      const schemaLocation = path.resolve(
        `${__dirname}/__data__/schema/tweet.json`
      );
      schema = JSON.parse(
        (await fs.readFile(schemaLocation)).toString()
      ) as unknown as DocumentNode;
    });

    it("should returned a parsed schema", () => {
      expect.hasAssertions();

      const parsed = parseSchema(schema);

      expect(parsed).toMatchSnapshot();
    });
  });
});
