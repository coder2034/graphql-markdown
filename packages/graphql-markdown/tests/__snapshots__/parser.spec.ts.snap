// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser parseSchema should returned a parsed schema 1`] = `
Array [
  Object {
    "directives": Array [
      Object {
        "arguments": Array [
          Object {
            "kind": "Argument",
            "name": "url",
            "value": "https://www.w3.org/TR/SRI",
          },
        ],
        "kind": "Directive",
        "name": "@specifiedBy",
      },
    ],
    "kind": "ScalarTypeDefinition",
    "name": "SRI",
    "simplifiedKind": "scalar",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "id",
        "type": "ID",
      },
    ],
    "interfaces": Array [],
    "kind": "InterfaceTypeDefinition",
    "name": "Node",
    "simplifiedKind": "InterfaceTypeDefinition",
  },
  Object {
    "directives": Array [],
    "kind": "EnumTypeDefinition",
    "name": "MediaType",
    "simplifiedKind": "enum",
    "values": Array [
      Object {
        "directives": Array [],
        "kind": "EnumValueDefinition",
        "name": "IMAGE",
      },
      Object {
        "directives": Array [],
        "kind": "EnumValueDefinition",
        "name": "VIDEO",
      },
    ],
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "url",
        "type": "Url",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "sri",
        "type": "SRI",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "type",
        "type": "MediaType",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Media",
    "simplifiedKind": "object",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "message",
        "type": "String",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Error",
    "simplifiedKind": "object",
  },
  Object {
    "directives": Array [],
    "kind": "UnionTypeDefinition",
    "name": "TweetResponse",
    "simplifiedKind": "union",
    "types": Array [
      Object {
        "isList": false,
        "isNull": true,
        "type": "Tweet",
      },
      Object {
        "isList": false,
        "isNull": true,
        "type": "Error",
      },
    ],
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "id",
        "type": "ID",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "body",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "date",
        "type": "Date",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "Author",
        "type": "User",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "Stats",
        "type": "Stat",
      },
      Object {
        "arguments": Array [
          Object {
            "defaultValue": undefined,
            "description": undefined,
            "directives": Array [],
            "isList": false,
            "isNull": true,
            "kind": "InputValueDefinition",
            "name": "meta",
            "type": "String",
          },
        ],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "getMeta",
        "type": "Meta",
      },
    ],
    "interfaces": Array [
      Object {
        "isList": false,
        "isNull": true,
        "type": "Node",
      },
    ],
    "kind": "ObjectTypeDefinition",
    "name": "Tweet",
    "simplifiedKind": "object",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "FieldDefinition",
        "name": "id",
        "type": "ID",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "username",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "first_name",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "last_name",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "full_name",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "name",
        "type": "String",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": "reason",
                "value": "Replaced by the \`Media\` field \`avatar\`.",
              },
            ],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "avatar_url",
        "type": "Url",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "avatar",
        "type": "Media",
      },
    ],
    "interfaces": Array [
      Object {
        "isList": false,
        "isNull": true,
        "type": "Node",
      },
    ],
    "kind": "ObjectTypeDefinition",
    "name": "User",
    "simplifiedKind": "object",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "views",
        "type": "Int",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": "reason",
                "value": "Use \`reactions\`.",
              },
            ],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "likes",
        "type": "Int",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "reactions",
        "type": "Int",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "retweets",
        "type": "Int",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "responses",
        "type": "Int",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Stat",
    "simplifiedKind": "object",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "id",
        "type": "ID",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "date",
        "type": "Date",
      },
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "type",
        "type": "String",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Notification",
    "simplifiedKind": "object",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "arguments": Array [],
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "FieldDefinition",
        "name": "count",
        "type": "Int",
      },
    ],
    "interfaces": Array [],
    "kind": "ObjectTypeDefinition",
    "name": "Meta",
    "simplifiedKind": "object",
  },
  Object {
    "directives": Array [],
    "kind": "ScalarTypeDefinition",
    "name": "Url",
    "simplifiedKind": "scalar",
  },
  Object {
    "directives": Array [],
    "kind": "ScalarTypeDefinition",
    "name": "Date",
    "simplifiedKind": "scalar",
  },
  Object {
    "arguments": Array [
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "InputValueDefinition",
        "name": "id",
        "type": "ID",
      },
    ],
    "description": undefined,
    "directives": Array [],
    "isList": false,
    "isNull": true,
    "kind": "OperationTypeDefinition",
    "name": "Tweet",
    "operation": "Query",
    "simplifiedKind": "query",
    "type": "TweetResponse",
  },
  Object {
    "arguments": Array [
      Object {
        "defaultValue": "5",
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "limit",
        "type": "Int",
      },
      Object {
        "defaultValue": "0",
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "skip",
        "type": "Int",
      },
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "sort_field",
        "type": "String",
      },
      Object {
        "defaultValue": "ASC",
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "sort_order",
        "type": "String",
      },
    ],
    "description": undefined,
    "directives": Array [],
    "isList": true,
    "isNull": false,
    "kind": "OperationTypeDefinition",
    "name": "Tweets",
    "operation": "Query",
    "simplifiedKind": "query",
    "type": "TweetResponse",
  },
  Object {
    "arguments": Array [],
    "description": undefined,
    "directives": Array [
      Object {
        "arguments": Array [],
        "kind": "Directive",
        "name": "@deprecated",
      },
    ],
    "isList": false,
    "isNull": true,
    "kind": "OperationTypeDefinition",
    "name": "TweetsMeta",
    "operation": "Query",
    "simplifiedKind": "query",
    "type": "Meta",
  },
  Object {
    "arguments": Array [
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "InputValueDefinition",
        "name": "id",
        "type": "ID",
      },
    ],
    "description": undefined,
    "directives": Array [
      Object {
        "arguments": Array [
          Object {
            "kind": "Argument",
            "name": "reason",
            "value": "Use \`Users\`.",
          },
        ],
        "kind": "Directive",
        "name": "@deprecated",
      },
    ],
    "isList": false,
    "isNull": true,
    "kind": "OperationTypeDefinition",
    "name": "User",
    "operation": "Query",
    "simplifiedKind": "query",
    "type": "User",
  },
  Object {
    "arguments": Array [
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": true,
        "isNull": false,
        "kind": "InputValueDefinition",
        "name": "id",
        "type": "ID",
      },
    ],
    "description": undefined,
    "directives": Array [],
    "isList": true,
    "isNull": false,
    "kind": "OperationTypeDefinition",
    "name": "Users",
    "operation": "Query",
    "simplifiedKind": "query",
    "type": "User",
  },
  Object {
    "arguments": Array [],
    "description": undefined,
    "directives": Array [
      Object {
        "arguments": Array [],
        "kind": "Directive",
        "name": "@deprecated",
      },
    ],
    "isList": false,
    "isNull": true,
    "kind": "OperationTypeDefinition",
    "name": "NotificationsMeta",
    "operation": "Query",
    "simplifiedKind": "query",
    "type": "Meta",
  },
  Object {
    "arguments": Array [
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": "reason",
                "value": "Use \`content\`.",
              },
            ],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "input",
        "type": "TweetContent",
      },
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "content",
        "type": "TweetContent",
      },
    ],
    "description": undefined,
    "directives": Array [],
    "isList": false,
    "isNull": true,
    "kind": "OperationTypeDefinition",
    "name": "createTweet",
    "operation": "Mutation",
    "simplifiedKind": "mutation",
    "type": "Tweet",
  },
  Object {
    "arguments": Array [
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "InputValueDefinition",
        "name": "id",
        "type": "ID",
      },
    ],
    "description": undefined,
    "directives": Array [],
    "isList": false,
    "isNull": true,
    "kind": "OperationTypeDefinition",
    "name": "deleteTweet",
    "operation": "Mutation",
    "simplifiedKind": "mutation",
    "type": "Tweet",
  },
  Object {
    "arguments": Array [
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": false,
        "kind": "InputValueDefinition",
        "name": "id",
        "type": "ID",
      },
    ],
    "description": undefined,
    "directives": Array [],
    "isList": false,
    "isNull": true,
    "kind": "OperationTypeDefinition",
    "name": "markTweetRead",
    "operation": "Mutation",
    "simplifiedKind": "mutation",
    "type": "Boolean",
  },
  Object {
    "directives": Array [],
    "fields": Array [
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [
          Object {
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": "reason",
                "value": "Use \`body\`.",
              },
            ],
            "kind": "Directive",
            "name": "@deprecated",
          },
        ],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "content",
        "type": "String",
      },
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "body",
        "type": "String",
      },
    ],
    "kind": "InputObjectTypeDefinition",
    "name": "TweetContent",
    "simplifiedKind": "input",
  },
  Object {
    "arguments": Array [
      Object {
        "defaultValue": undefined,
        "description": undefined,
        "directives": Array [],
        "isList": false,
        "isNull": true,
        "kind": "InputValueDefinition",
        "name": "limit",
        "type": "Int",
      },
    ],
    "description": undefined,
    "directives": Array [],
    "isList": true,
    "isNull": false,
    "kind": "OperationTypeDefinition",
    "name": "Notifications",
    "operation": "Subscription",
    "simplifiedKind": "subscription",
    "type": "Notification",
  },
]
`;
